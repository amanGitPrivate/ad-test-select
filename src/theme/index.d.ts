export interface themeInterface {
  size: {
    small: string;
    large: string;
  };
  weight: {
    regular: string;
    medium: string;
    bold: string;
  };
  colors: {
    100: string;
    200: string;
    300: string;
    400: string;
    500: string;
    600: string;
    700: string;
  };
  spacing: (spacing: number) => string;
  shadow: string;
}
