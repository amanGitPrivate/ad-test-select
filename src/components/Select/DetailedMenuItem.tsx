import { AdditionalInfoChilren } from "./Select.interface";
import { StyledAdditonalInfoOption } from "./Select.styled";

const DetailedMenuItem = ({ option }: { option: AdditionalInfoChilren }) => (
  <div key={option.label}>
    <div>{option.label}</div>
    <StyledAdditonalInfoOption>
      {option.additionalInfo}
    </StyledAdditonalInfoOption>
  </div>
);

export default DetailedMenuItem;
