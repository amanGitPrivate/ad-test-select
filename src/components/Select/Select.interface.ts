/* children interace */

export interface ListChildren {
  value: string;
  label: string;
  additionalInfo?: string
}

interface ListOption {
  categoryName: string;
  children: ListChildren[];
}

export interface AdditionalInfoChilren {
  value: string;
  label: string;
  additionalInfo: string
}
export interface Selected {
  label?: string;
  value: string;
  addintionalInfo?: string;
}

export interface SelectProps {
  menuOptions: ListOption[];
  searchable?: boolean;
  searchPlaceHolder?: string;
  onChange: (option: ListChildren) => void;
  selected?: {
    value: string;
    label?: string;
  };
  renderOption?: (option: ListChildren) => void;
  isMenuOpen?: boolean;
  dropDownlabel?: string;
  dropDownPlaceholder?: string;
}

