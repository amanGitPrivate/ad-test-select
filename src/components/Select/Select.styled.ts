import styled from "styled-components";

import { themeInterface } from "../../theme/index.d";

export const StyledWrapper = styled.div`
  ${({ theme: { spacing, colors } }) => `
  height: ${spacing(8)};
  min-width: ${spacing(40)};
  border: 1px solid ${colors[500]};
  border-radius: ${spacing(1)};
  &:focus {
    outline: none
  }
  `}
`;

export const StyledTitleText = styled.div`
  margin: ${(props) => `${props.theme.spacing(1)} ${props.theme.spacing(3)}`};
  margin-right: 30px;
  font-weight: 300;
`;

export const StlyedHeaderTitle = styled.div`
  font-weight: 300;
`;

export const StyledDropdownButton = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: relative;
  background-color: transparent;
  cursor: pointer;
  height: 100%;
  width: 100%;
`;

export const StyledArrow = styled.span`
  margin-right: 8px;
  transform: ${(props: { down?: boolean }) =>
    props.down ? "rotate(0deg)" : "rotate(180deg)"};
  transition: transform 0.25s linear;
`;

export const StyledOption = styled.div`
  cursor: pointer;
  align-items: center;
  ${({
    theme: { spacing, colors },
    selected,
    hidden,
    navigated,
  }: {
    selected?: boolean;
    navigated?: boolean;
    theme: themeInterface;
    hidden: boolean;
  }) => `
    padding: ${spacing(1)} ${spacing(3)};
    background-color:  ${
      selected ? colors[100] : navigated ? colors[400] : "transparent"
    };
    &:hover {
      background-color: ${selected ? colors[200] : colors[400]};
    }
    color: ${selected ? colors[300] : "black"};
    margin: ${spacing(1)} 0};
    min-height: ${spacing(6)};
    max-height: ${spacing(15)};
    display: ${hidden ? "none" : "flex"}
  `}
`;

export const StyledAdditonalInfoOption = styled.div`
  ${({ theme: { size } }) => `
    font-size: ${size.small};
  `}
`;

export const StyledOptions = styled.div`
  overflow: auto;
  max-height: 215px;
  margin-top: 4px;
  z-index: 10;
  position: relative;
  ${({ theme: { spacing, colors, shadow } }) => `
    border: 1px solid ${colors[500]};
    border-radius: ${spacing(1)};
    box-shadow: ${shadow};
    background-color: ${colors[300]};
  `}
`;

export const SelectedIcon = styled.span`
  visibility: ${({ selected }: { selected?: boolean }) =>
    selected ? `visible` : `hidden`};
  margin: 0 12px 0 0;
`;

export const StyledSelectImages = styled.img`
  ${({ theme: { spacing } }) => `
    height:  ${spacing(4)};
    width:  ${spacing(4)};
  `}
`;

export const StyledInputSearch = styled.input`
  ${({ theme: { spacing, colors } }) => `
    border-radius: ${spacing(1)};
    &:focus {
      outline: none
    }
    width: 99%;
    margin-top: ${spacing(1)};
    height: ${spacing(10)};
    padding-left: ${spacing(3)};
    border: 1px solid ${colors[500]}
  `}
`;

export const SelectGroup = styled.div`
  border-bottom: ${({ theme: { colors } }) => `2px solid ${colors[500]}`};
`;

export const StyledLabel = styled.div`
  ${({ theme: { colors, spacing } }) => `
    color: ${colors[700]};
    margin-bottom: ${spacing(1.5)}
  `}
`;

export const StyledDropDownPlaceholder = styled.span`
  ${({ theme: { colors } }) => `
    color: ${colors[600]};
  `}
`;

export const StyledNoOption = styled.div`
  padding: ${({ theme: { spacing } }) => `${spacing(1)}`};
`;
