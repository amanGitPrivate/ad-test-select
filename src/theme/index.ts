export const theme = {
  size: {
    small: "12px",
    large: "16px",
  },
  weight: {
    regular: "regular",
    medium: "medium",
    bold: "bold",
  },
  colors: {
    100: "#00A3BE",
    200: "#05778F",
    300: "#F9FAFC",
    400: "#ECEEF4",
    500: "#CDD0E0",
    600: "#6E7492",
    700: "#191D2F",
  },
  spacing: (spacing: number) => {
    return `${4 * spacing}px`;
  },
  shadow: `0 2px 8px 0 #6E749226`,
};
