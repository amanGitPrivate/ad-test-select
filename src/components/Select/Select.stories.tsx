import { ComponentStory, ComponentMeta } from "@storybook/react";

import Select from "./Select";

export default {
  title: "Components/Select",
  component: Select,
} as ComponentMeta<typeof Select>;

const Template: ComponentStory<typeof Select> = (args) => <Select {...args} />;

export const Primary = Template.bind({});

Primary.args = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Hopper",
          value: "fp1",
        },
        {
          label: "Holberton",
          value: "fp2",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Anotelli",
          value: "sp1",
        },
        {
          label: "Bartik",
          value: "sp2",
        },
        {
          label: "Teitelbaum",
          value: "sp3",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: false,
  searchPlaceHolder: "Search Here !!",
  dropDownlabel: "My Dropdown !!",
};

export const Searchable = Template.bind({});
Searchable.args = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Hopper",
          value: "fp1",
        },
        {
          label: "Holberton",
          value: "fp2",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Anotelli",
          value: "sp1",
        },
        {
          label: "Bartik",
          value: "sp2",
        },
        {
          label: "Teitelbaum",
          value: "sp3",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: true,
  dropDownPlaceholder: "Your options",
  searchPlaceHolder: "Search Here !!",
};

export const DefaultValue = Template.bind({});
DefaultValue.args = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Hopper",
          value: "fp1",
        },
        {
          label: "Holberton",
          value: "fp2",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Anotelli",
          value: "sp1",
        },
        {
          label: "Bartik",
          value: "sp2",
          additionalInfo:
          "Grace Hopper was an American computer scientist and US Navy admiral.",
        },
        {
          label: "Teitelbaum",
          value: "sp3",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: false,
  selected: {
    value: "fp2",
  },
  searchPlaceHolder: "Search Here !!",
};

export const AdditionalInfo = Template.bind({});
AdditionalInfo.args = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Hopper",
          value: "fp1",
          additionalInfo:
            "Grace Hopper was an American computer scientist and US Navy admiral.",
        },
        {
          label: "Holberton",
          value: "fp2",
          additionalInfo:
            "Frances Elizabeth Holberton was one of the first computer programmer",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Teitelbaum",
          value: "sp1",
          additionalInfo:
            "Ruth Teitelbaum was one of the first computer programmer",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: false,
  dropDownPlaceholder: "Additional search options",
  searchPlaceHolder: "Search Here !!",
};

export const DefaultOpen = Template.bind({});
DefaultOpen.args = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Hopper",
          value: "fp1",
        },
        {
          label: "Holberton",
          value: "fp2",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Anotelli",
          value: "sp1",
        },
        {
          label: "Bartik",
          value: "sp2",
        },
        {
          label: "Teitelbaum",
          value: "sp3",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: false,
  isMenuOpen: true,
};
