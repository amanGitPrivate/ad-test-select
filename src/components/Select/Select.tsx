import { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from "react";
import { createGlobalStyle, ThemeProvider } from "styled-components";
import DetailedMenuItem from "./DetailedMenuItem";
import {
  StyledWrapper,
  StlyedHeaderTitle,
  StyledDropdownButton,
  StyledArrow,
  StyledOption,
  StyledOptions,
  StyledTitleText,
  SelectedIcon,
  StyledSelectImages,
  StyledInputSearch,
  SelectGroup,
  StyledLabel,
  StyledDropDownPlaceholder,
  StyledNoOption,
} from "./Select.styled";

import ArrowUp from "../../assets/chevron-down.svg";
import ArrowDown from "../../assets/chevron-down.svg";
import Check from "../../assets/check-light.svg";
import {
  isNavigationKey,
  getNavigationAction,
  NAVIGATION,
} from "../../helpers/keyboard";

import { SelectProps, Selected, ListChildren, AdditionalInfoChilren } from "./Select.interface";
import { theme } from "../../theme";

const StyledGlobal = createGlobalStyle`
  body {
    font-family: -apple-sysyem, Roboto, "Segoe UI", Arial, sans-serif;
  }
`;

const Select = ({
  isMenuOpen = false,
  menuOptions,
  selected,
  searchable,
  searchPlaceHolder,
  renderOption,
  onChange,
  dropDownlabel,
  dropDownPlaceholder,
}: SelectProps) => {
  const [menuOpen, toggleMenu] = useState(isMenuOpen);
  const [selectedItem, setSelectedItem] = useState(selected);
  const [keyword, updateKeyWord] = useState("");
  const [navigatedOptionIndex, updateNavigationOptionIndex] = useState(-1);
  const [navigatedOption, updateNavigationOption] = useState<ListChildren>({
    value: "",
    label: "",
  });

  const textRef = useRef<HTMLInputElement | null>(null);
  const node = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (selected) {
      selectOption(selected);
    }
    // add when mounted
    document.addEventListener("mousedown", handleClick);
    // return function to be called when unmounted
    return () => {
      document.removeEventListener("mousedown", handleClick);
    };
  }, []);

  const handleClick = (e: any) => {
    if (node.current) {
      if (node.current.contains(e.target)) {
        return;
      }
    }
    updateKeyWord(""); // outside click
    toggleMenu(false); // outside click
  };

  const selectOption = (item: Selected) => {
    let selectedItem: any = [];
    for (let i = 0; i < menuOptions.length; i++) {
      selectedItem = menuOptions[i].children.filter(
        (child) => child.value === item.value
      );
      if (selectedItem) {
        break;
      }
    }
    if (selectedItem.length) {
      setSelectedItem(selectedItem[0]);
    }
  };

  const selectItem = (item: ListChildren) => {
    setSelectedItem(item);
    onChange(item);
    toggleMenu(false);
    updateNavigationOptionIndex(-1);
    updateNavigationOption({
      value: "",
      label: "",
    });
    updateKeyWord("");
  };

  const filterList = (e: ChangeEvent<HTMLInputElement>) => {
    updateKeyWord(e.target.value.toLowerCase());
  };

  const renderListOption = (option: ListChildren) => {
    if (renderOption && option) {
      return renderOption(option);
    }
    if (option.additionalInfo) {
      return <DetailedMenuItem option={option as AdditionalInfoChilren} />;
    }
    return <span>{option.label}</span>;
  };

  const listItems = () => {
    let tempList = [...menuOptions];
    const selectedItemValue = selectedItem?.value;

    if (tempList.length) {
      return tempList.map((item) => (
        <SelectGroup key={item.categoryName}>
          {item.children.map((child) => (
            <StyledOption
              key={child.label}
              navigated={navigatedOption.value === child.value}
              selected={child.value === selectedItemValue}
              hidden={
                !child.label.toLowerCase().includes(keyword.toLowerCase())
              }
              onClick={() => selectItem(child)}
            >
              <SelectedIcon selected={child.value === selectedItemValue}>
                {<StyledSelectImages src={Check} />}
              </SelectedIcon>
              {renderListOption(child)}
            </StyledOption>
          ))}
        </SelectGroup>
      ));
    }
    return <StyledNoOption>No Options avialable</StyledNoOption>;
  };

  const handleNavigation = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.defaultPrevented || !isNavigationKey(event.key)) {
      return;
    }
    event.preventDefault();
    const navigationAction = getNavigationAction(event.key) || "";

    let menuLength = 0;
    let menuCollection = [];
    for (let i = 0; i < menuOptions.length; i++) {
      menuLength += menuOptions[i].children.length;
      menuCollection.push(...menuOptions[i].children);
    }
    const optionCount = menuOptions ? menuLength || 1 : 0;

    if (optionCount < 1) {
      return;
    }

    switch (navigationAction) {
      case NAVIGATION.CLOSE: {
        toggleMenu(false);
        break;
      }
      case NAVIGATION.SELECT: {
        selectItem(navigatedOption);
        break;
      }

      case NAVIGATION.UP:
      case NAVIGATION.DOWN: {
        let nextIndex =
          navigatedOptionIndex + (navigationAction === NAVIGATION.UP ? -1 : 1);
        nextIndex = nextIndex < 0 ? optionCount - 1 : nextIndex;
        nextIndex = nextIndex >= optionCount ? 0 : nextIndex;
        updateNavigationOptionIndex(nextIndex);
        updateNavigationOption(menuCollection[nextIndex]);
      }
    }
  };

  return (
    <ThemeProvider theme={theme}>
      <StyledLabel>{dropDownlabel}</StyledLabel>
      <StyledWrapper ref={node} onKeyDown={handleNavigation} tabIndex={0}>
        <StyledDropdownButton onClick={() => toggleMenu(true)}>
          <StlyedHeaderTitle>
            <StyledTitleText>
              {selectedItem ? (
                selectedItem.label
              ) : (
                <StyledDropDownPlaceholder>
                  {dropDownPlaceholder}
                </StyledDropDownPlaceholder>
              )}
            </StyledTitleText>
          </StlyedHeaderTitle>
          {menuOpen ? (
            <StyledArrow>{<StyledSelectImages src={ArrowUp} />}</StyledArrow>
          ) : (
            <StyledArrow down>
              {<StyledSelectImages src={ArrowDown} />}
            </StyledArrow>
          )}
        </StyledDropdownButton>
        {menuOpen && (
          <div>
            {searchable && (
              <StyledInputSearch
                ref={textRef}
                placeholder={searchPlaceHolder}
                onClick={(e) => e.stopPropagation()}
                onChange={(e) => filterList(e)}
              />
            )}
            <StyledOptions tabIndex={0}>{listItems()}</StyledOptions>
          </div>
        )}
      </StyledWrapper>
      <StyledGlobal />
    </ThemeProvider>
  );
};

export default Select;
