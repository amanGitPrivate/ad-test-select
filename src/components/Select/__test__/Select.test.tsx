import { mount } from "enzyme";
import { ThemeProvider } from "styled-components";
import {
  StyledOptions,
  StyledDropdownButton,
  StyledTitleText,
  StyledOption,
} from "../Select.styled";

import Select from "../Select";
import { theme } from "../../../theme";

import "jsdom-global/register";

const props = {
  menuOptions: [
    {
      categoryName: "firstPerson",
      children: [
        {
          label: "Option 1",
          value: "fp1",
        },
        {
          label: "Option 2",
          value: "fp2",
          additionalInfo: "ImportantInfo here !!",
        },
        {
          label: "Option 3",
          value: "fp3",
        },
      ],
    },
    {
      categoryName: "secondPerson",
      children: [
        {
          label: "Option 1",
          value: "sp1",
        },
        {
          label: "Option 2",
          value: "sp2",
          additionalInfo: "ImportantInfo here !!",
        },
        {
          label: "Option 3",
          value: "sp3",
        },
      ],
    },
  ],
  onChange: () => {},
  searchable: false,
  searchPlaceHolder: "Search Here !!",
};

it("Should render with dropdown", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select {...props} />
    </ThemeProvider>
  );
  expect(select).toMatchSnapshot();
});

it("Should render with dropdown open", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select {...props} isMenuOpen={true} />
    </ThemeProvider>
  );
  expect(select.find(StyledOptions).exists()).toBe(true);
});

it("Should open dropdown on click", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select {...props} />
    </ThemeProvider>
  );
  select.find(StyledDropdownButton).simulate("click");
  expect(select.find(StyledOptions).exists()).toBe(true);
});

it("Should select the correct item on click", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select {...props} />
    </ThemeProvider>
  );
  select.find(StyledDropdownButton).simulate("click");
  select.find(StyledOption).at(0).simulate("click");
  expect(select.find(StyledTitleText).text().length).toBeGreaterThan(0);
});

it("Should select the sent option", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select
        {...props}
        selected={{
          value: "fp2",
        }}
      />
    </ThemeProvider>
  );
  expect(select.find(StyledTitleText).text()).toEqual("Option 2");
});

it("Should render the extended menu with additional item using renderOption", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select
        {...props}
        selected={{
          value: "fp2",
        }}
        renderOption={(option) => (
          <div key={option.label}>
            <div>{option.label}</div>
            <div>{option.additionalInfo}</div>
          </div>
        )}
      />
    </ThemeProvider>
  );
  select.find(StyledDropdownButton).simulate("click");
  expect(
    select.find(StyledOption).at(1).text().indexOf("ImportantInfo here !!")
  ).toBeGreaterThan(-1)
});

it("Should render the extended menu with additional item using list", () => {
  const select = mount(
    <ThemeProvider theme={theme}>
      <Select
        {...props}
        selected={{
          value: "fp2",
        }}
      />
    </ThemeProvider>
  );
  select.find(StyledDropdownButton).simulate("click");
  expect(
    select.find(StyledOption).at(1).text().indexOf("ImportantInfo here !!")
  ).toBeGreaterThan(-1)
});
