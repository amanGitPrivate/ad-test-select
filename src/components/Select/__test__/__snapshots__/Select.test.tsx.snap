// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render with dropdown 1`] = `ReactWrapper {}`;
